### PR TITLE
DimShuffle layer

### PR DESCRIPTION
**Pull request by Britefury**
Implemented DimShuffleLayer in lasagne.layers.shape. Implemented tests.

### CODE REVIEW

**Comment by f0k** [comment]
Thanks, looks good superficially (didn't check the implementation in detail)! It's just missing a docstring.

File: lasagne/layers/shape.py [hunk 1]
```diff
@@ -139,6 +139,62 @@ def get_output_for(self, input, *args, **kwargs):
...
+            elif p == 'x':
+                # Broadcast; will be of size 1
+                o = 1
+            else:
+                raise RuntimeError("invalid pattern entry, should have "
+                    "caught in the constructor")
+            output_shape.append(o)
+
+        for i, (dim_size, used) in enumerate(zip(input_shape, dims_used)):
+            if not used and dim_size != 1:
```
Comment (f0k) [inline lasagne/layers/shape.py]: `dim_size` could also be `None`. In this case, we cannot know whether it's okay to collapse it and should just assume that it is.

**Comment by benanne** [comment]
Looks great! The tests look good as well. I have just a few minor comments:

* Since the Theano function this wraps is called `dimshuffle` and not `dim_shuffle`, I think it would make more sense to call this layer `DimshuffleLayer` (lowercase s). Seems a bit more consistent that way. I don't know, I guess I have API-OCD :p 

* To support layer naming (and possibly additional features in the future), make sure to add `**kwargs` to `__init__()` and pass it to the superclass's `__init__()` method.

**Comment by f0k** [comment]
> To support layer naming (and possibly additional features in the future), [...]

Ah, that's missing in `ReshapeLayer` as well... Sander, could you fix this directly on master, please?

**Comment by benanne** [comment]
will do!

**Comment by Britefury** [comment]
Should all be done now (`**kwargs`, `DimshuffleLayer` and `None` in input shape).

**Comment by benanne** [comment]
Awesome! Will merge this tomorrow if there are no further comments.

**Comment by f0k** [comment]
It's still missing the docstring! It should have a class docstring explaining what the purpose of the layer is and what the constructor arguments mean. (Although I'm not sure if the latter shouldn't be in the constructor docstring instead? I've seen both here in Lasagne.)

**Comment by benanne** [comment]
Whoops, good catch! Yeah, we need to sort that out. What's the preferred way to do it? Are there any advantages/disadvantages to putting it in the class / constructor docstring?

Also, this is probably another `Layer` that deserves a shortcut, because people typically don't see a shuffling operation as a separate layer. Since we already have `lasagne.layers.reshape` and such, could you add the following line below the class definition to create a shortcut for it?

```
dimshuffle = DimshuffleLayer
```

Thanks!

**Comment by craffel** [comment]
My impression/convention is that the class docstring should document the class and its members, and the `__init__` docstring should document the init method and its parameters.

**Comment by benanne** [comment]
But in practice, where do you draw the line? For example, sometimes it's hard to describe what a layer does without referring to some of its parameters by name. I don't think it's always that clear-cut what should go where.

**Comment by craffel** [comment]
> But in practice, where do you draw the line? For example, sometimes it's hard to describe what a layer does without referring to some of its parameters by name.

The line is members (i.e. `foo.b`) vs. initialization parameters (i.e. `foo = Foo(b)`).  For example, the class docstring for `lasagne.layers.Layer` should descibe what `input_layer` and `input_shape` are, while the docstring for its initializer should describe what `incoming` is.  Or, in a `DenseLayer`, the class docstring for the member `W` should describe what a weight matrix is, while the initializer docstring for `W` should describe the suitable methods for initializing it.  See https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#documenting-classes for some more info.

**Comment by Britefury** [comment]
The doc-string and shortcut are done. I also added `DimshuffleLayer` and the shortcut to `__all__`, something else I omitted! :)

**Comment by benanne** [comment]
@craffel: that makes sense, thanks. We should keep this in mind for the future (gonna tag #26 here so this shows up there).

@Britefury: looks great, thanks! Second attempt:I'll merge this tomorrow if there are no further comments.

**Comment by benanne** [comment]
Merging, thanks!

**Pull request merged**